server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -648,6 +648,7 @@ public Iterator<Setting<?>> settings() {
     private final Double writeLoadForecast;
     @Nullable
     private final Long shardSizeInBytesForecast;
+    private final boolean isDataStream;
```

Comment:
I remember in the past there was push back for also recording in `IndexMetadata` that an index is part of a data stream. Right now this can be derived from `DataStream`. Indices can be moved out of a data stream and then this needs to be updated as well. 


Thinking more about this, I think we don't need to this change. I think we should make the decision on whether to use the new uuid generator based on index mode only.

Reply:
Do we know the reason behind the push back? Is it because `IndexMetadata` is more about indices rather then data streams? Or maybe because the status would not reflect reality in case the index is taken out of the data stream?

Reply:
Yes, because we would need to maintain in two places whether an index is part of a data stream. In `DataStream` itself and in `IndexMetadata`.  And this increases the chances of potential bugs.

For example you need to think about restoring only a backing index from a snapshot or the modify data stream api. The changes in the PR didn't take this into account, because according to the `isDataStream` property an index would be part of a data stream even if no data stream referred to this index.